// @scaffold-digital/ionic-hardware-buttons/dist/providers/hardware-buttons.js
import { Injectable } from '@angular/core';
import { App, Platform } from 'ionic-angular';
var callbacks = {};
/**
 * Defines a method that will be called when the hardware back button is pressed
 * and this view is visible. Returning false will prevent the default
 * functionality.
 */
export function BackButton() {
    return function (target, propertyKey, descriptor) {
        var constructor = target.constructor;
        callbacks[constructor] = propertyKey;
    };
}
var HardwareButtons = (function () {
    function HardwareButtons(app, platform) {
        this.app = app;
        this.platform = platform;
        this.viewStack = [];
    }
    HardwareButtons.prototype.getCallbackForViewCtrl = function (viewCtrl) {
        var instance = viewCtrl.instance;
        var component = instance.constructor;
        var method = callbacks[component];
        if (!method)
            return null;
        var callback = instance[method];
        return callback;
    };
    HardwareButtons.prototype.getCurrentViewCtrl = function () {
        return this.viewStack.length > 0 ? this.viewStack[this.viewStack.length - 1] : null;
    };
    HardwareButtons.prototype.init = function () {
        var _this = this;
        var app = this.app;
        var platform = this.platform;
        app.viewDidEnter.subscribe(function (viewCtrl) {
            _this.viewStack.push(viewCtrl);
        });
        app.viewDidLeave.subscribe(function (viewCtrl) {
            var index = _this.viewStack.lastIndexOf(viewCtrl);
            _this.viewStack.splice(index, 1);
        });
        platform.registerBackButtonAction(function () {
            var viewCtrl = _this.getCurrentViewCtrl();
            if (!viewCtrl)
                return _this.app.goBack();
            var result = _this.triggerCallbackForViewCtrl(viewCtrl);
            if (result === false)
                return;
            app.goBack();
        });
    };
    HardwareButtons.prototype.triggerCallbackForViewCtrl = function (viewCtrl) {
        var callback = this.getCallbackForViewCtrl(viewCtrl);
        if (!callback)
            return;
        return callback();
    };
    HardwareButtons.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    HardwareButtons.ctorParameters = function () { return [
        { type: App, },
        { type: Platform, },
    ]; };
    return HardwareButtons;
}());
export { HardwareButtons };
